Dev Queue pull request to master

mcdonals opened the pull request:
Pull request to merge dev_queue into master. At this point Queue supports enque(), deque() and a max capacity. The QueueNode class supports the generic type signature for it's elements. 

JUnit classes created for both Queue class and QueueNode class and they provide 100% code coverage.

mcdonals committed:
Initial class files and test files for the Queue data structure
implementation. QueueNode's constructor is finished and has been
validated by QueueNodeTest.

mcdonals committed:
Added enque, deque, getCapacity and getSize methods to the Queue class.

mcdonals committed:
Added getElement(), getNextNode() and setNextNode() methods to 
QueueNode class

mcdonals committed:
Added testQueueNodeSetNextNode() to reach 100% code coverage.

mcdonals committed:
Added a new test method to test both enque() and deque()

mcdonals committed:
Final commit for Queue basic functionality. Supports Queue enque() and
deque() with a Queue capacity and QueueNodes that accept any generic
type. QueueNodeTest and QueueTest provide 100% test coverage for both
new classes.

jonjoncronin commented on src/queue/Queue.java:
```diff
@@ -0,0 +1,116 @@
+    private int capacity;
+    private int size;
+    
+    // Define the return values the Queue class methods may use
```
Ya - Enums? why was that a problem?

mcdonals replied:
Basically my JUnit skills are still weak and enums in Java are actually mini classes. When i tried it as an enum type there were some functions that i needed to overload and test or jacoco wouldn't give me 100% test coverage anymore. So I abandoned the enum class type error return method for the time being. It's an enhancement I could look at adding again in the future.

jonjoncronin commented on src/queue/Queue.java:
```diff
@@ -0,0 +1,116 @@
+     * Default constructor. Sets head and tail to null, max capacity to 10 and
+     * current size to 0.
+     */
+    public Queue() {
```
do we have a coding standard? do you really care if our code looks a little different?

mcdonals replied:
Yeah I was wondering if we should try and care. I'm fine either way. I've been trying to follow what's in "Core Java for the Impatient" which is on Safari. I think it's closer to the normal Java convention than the typical C convention, but I don't really mind either way.

jonjoncronin commented on src/queue/Queue.java:
```diff
@@ -0,0 +1,116 @@
+     * @param capacity 
+     */
+    public Queue(int capacity) {
+        this();
```
I'm unfamiliar with this syntax? what does this(); line do?

mcdonals replied:
that calls the default Queue constructor from the overloaded constructor so you don't have to rewrite all the lines and only have to worry about the differences between the constructor calls. You can do the same thing with inherited classes and their constructors but then you use super() instead of this(). this is shorthand for this object instance, so from the object instance you are accessing the object instance.

jonjoncronin commented on src/queue/Queue.java:
```diff
@@ -0,0 +1,116 @@
+    }
+    
+    /**
+     * Returns a new node at the tail of the queue as long as their is space
```
...there is space...

mcdonals replied:
one would hope :)

jonjoncronin commented on src/queue/Queue.java:
```diff
@@ -0,0 +1,116 @@
+            head = node;
+            tail = node;
+        } else {
+            tail.setNextNode(node);
```
Should you check that node.next is also null? You don't want someone creating a cycle on the queue.

mcdonals replied:
Not a bad idea. Added the check and a new error case: Q_ERR_NEXT_NODE. Added corresponding JUnit case too.

jonjoncronin commented on src/queue/Queue.java:
```diff
@@ -0,0 +1,116 @@
+     * @return int capacity 
+     */
+    public int getCapacity() {
+        return capacity;
```
I don't remember - do you need to make this line "return this.capacity" or is "this." redundant?

mcdonals replied:
It's redundant, but there are times where you have conflict between a local variable and an instance variable with the same name. For example the following wouldn't work:

public void setCapacity(int capacity) {
    capacity = capacity;
}

but this would
public void setCapacity(int capacity) {
    this.capacity = capacity;
}

jonjoncronin commented on src/queue/QueueNode.java:
```diff
@@ -0,0 +1,51 @@
+ * 
+ * @author Scott McDonald and Jon Cronin
+ */
+public class QueueNode<E> {
```
What is this <> notation?

mcdonals replied:
This is the java generic type notation. You can have as many generic types in a class as you need. Users can then pass in any object for that type. It is similar to what you did using "object" in linkedList, but different because you can do some type checking on the generic types at runtime. I think it is the preferred style for Java API classes. The collections classes like ArrayList all use it. Core Java for the Impatient covers them well in chapter 6 "Generic Programming".

jonjoncronin commented on src/queue/QueueNode.java:
```diff
@@ -0,0 +1,51 @@
+    public void setNextNode(QueueNode next) {
+        this.nextNode = next;
+    }
+}
```
Where is the toString() and equals()?

mcdonals replied:
I'll add them in round two.

jonjoncronin commented on src/queue/Queue.java:
```diff
@@ -0,0 +1,116 @@
+    public int getSize() {
+        return size;
+    }
+    
```
Where is the toString() and equals()?

mcdonals replied:
Round two :)

jonjoncronin commented on test/queue/QueueTest.java:
```diff
@@ -0,0 +1,103 @@
+ */
+public class QueueTest extends TestCase {
+    
+    // Default JUnit constructor created by Netbeans. Leaving for now.
```
I tried using the automatic template that Netbeans creates as well for the stack implementation - I ended up modifying the code to mimic the test files we've already got.

jonjoncronin commented on test/queue/QueueTest.java:
```diff
@@ -0,0 +1,103 @@
+    public void testQueue() {
+        Queue myQ = new Queue();
+        Queue myQSize = new Queue(20);
+        assertTrue(myQ.deque() == null);
```
A pattern I've been following is to not include a function call in the assert(). This allows me to debug the test a little more easily as the result of the function is captured as a local variable - I can see what the value is a little better.

mcdonals replied:
Not a bad idea. I've updated all the asserts i had that made a function call directly in the assert.

jonjoncronin commented on test/queue/QueueTest.java:
```diff
@@ -0,0 +1,103 @@
+        // We have a default Q, so the capacity is 10, validate we can enqueue
+        // 10 nodes and that the 11th node will return Q_ERR_FULL
+        QueueNode<Integer> node;
+        for (int ii=0; ii<11; ii++) {
```
Just a question about your boundary value of 11. Is there a reason you pick < 11 rather than <= 10?

mcdonals replied:
I just think it's easier to read, but that's just me :)

jonjoncronin commented:
lots of questions about syntax that I need you to teach me but nothing looks blatantly wrong.

mcdonals committed:
Updated enque() so it validates that nextNode is null.

mcdonals committed:
Updated all asserts so they check local variables, not the direct return
values from the methods under test. This makes debugging failing tests a
little easier.
